README.txt
= thin_ham

Source: 
* http://github.com/kschiess/thin_ham

== DESCRIPTION:

thin_ham is a small tool to allow for fast prototyping of web sites. It will
publish files written in haml & sass [1] or alternatively just publish
whatever other files you happen to have lying around in the relevant
directories.

Since it is written as a web server, it will allow running javascript in your
prototype to use AJAX to load missing resources. This can be very useful when
developing modern websites. 

Note that thin_ham should under NO circumstances be used as a public web
server. It is not made that way. 


== SYNOPSIS:

Move to a directory of your choice (where your prototype will be located) and 
run: 

  thin_ham
  
thin_ham will then output an url under which the current directory can be 
accessed locally. 

== LICENSE:

(The MIT License)

Copyright (c) 2009 Kaspar Schiess

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 1.0.0 / 2009-07-16

* Releasing thin_ham

spec/spec_helper.rb
require 'spec'

$:.unshift File.dirname(__FILE__) + "/../lib"

require 'thin_ham/init'

Spec::Runner.configure do |config| 
  config.mock_with :flexmock 
end

FIXTURES_ROOT = File.join(
  File.dirname(__FILE__), 
  '/fixtures')
  
# Generates a full path from a path relative to the fixtures root
#
def fixture(path)
  File.join(FIXTURES_ROOT, path)
end

def anything
  FlexMock.any
end

test/test_thin_ham.rb
require "test/unit"
require "thin_ham"

class TestThinHam < Test::Unit::TestCase
  def test_sanity
    flunk "write tests or I will kneecap you"
  end
end


spec/unit/hamlr_spec.rb
require File.dirname(__FILE__) + '/../spec_helper'

require 'thin_ham/hamlr'

require 'rack/mock'
require 'rack/lint'

describe ThinHam::Hamlr do
  it "should be initialized with a path" do
    ThinHam::Hamlr.new('.')
  end
  
  context 'fixtures (unit)' do
    attr_reader :hamlr
    before(:each) do
      @hamlr = ThinHam::Hamlr.new(
        FIXTURES_ROOT)
    end

    describe "#call" do
      before(:each) do
        # Don't duplicate instance - this makes mocking/stubbing possible
        flexmock(hamlr).should_receive(:dup).and_return(hamlr)
      end
    end
    describe "#http_answer" do
      attr_reader :mtime_http, :result
      before(:each) do
        @result = hamlr.http_answer(
          File.read(FIXTURES_ROOT + "/html_file.html"), 
          'text/html')
      end
      it "should return http code 200 OK" do
        result.first.should == 200
      end
      it "should have an header hash at second position" do
        result.at(1).should be_an_instance_of(Hash)
      end
      it "should have a body at last position" do
        result.last.size.should > 10
      end

      describe "header hash" do
        attr_reader :headers
        before(:each) do
          @headers = result.at(1)
        end
        it { headers.should have_key('Content-Length') }
        it { headers.should have_key('Content-Type') }
      end
    end
  end
  context 'fixtures (functional)' do
    attr_reader :request
    before(:each) do
      @request = Rack::MockRequest.new(
        Rack::Lint.new(
          ThinHam::Hamlr.new(
            FIXTURES_ROOT)))
    end
    it "should return a 404 if the file cannot be found" do
      request.get('/this_file_doesnt_exist').should be_not_found
    end
    it "should return 200 OK for an html file" do
      request.get('/html_file.html').should be_ok
    end
    it "should return the file content of html_file.html" do
      request.get('/html_file.html').should match(/a simple html file/)
    end
    
    context 'haml rendering' do
      it "should return 200 OK for test1.html" do
        request.get('/test1.html').should be_ok
      end
      it "should contain html header for test1.html" do
        request.get('/test1.html').should match(%r{<body>\s+This is a test\s+</body>})
      end
    end
    context 'sass rendering' do
      it "should return 200 OK for test1.css" do
        request.get('/test1.css').should be_ok
      end
    end
  end
end


lib/thin_ham/processors/haml.rb

require 'haml'

require 'thin_ham/processors/markup'

class ThinHam::Processors::Haml < ThinHam::Processors::Markup
  
  # Helper class for markup evaluation. 
  #
  class Helper
    attr_reader :base_path, :haml_processor
    def initialize(base_path, haml_processor)
      @base_path, @haml_processor = base_path, haml_processor
    end
    
    def include(file)
      include_name = File.join(base_path, file)
      haml_processor.serve(include_name)
    end
    
    def with_layout(file, &block)
      include_name = File.join(base_path, file)
      haml_processor.serve(include_name) do 
        capture_haml(&block)
      end
    end
  end
  
  # Serves the file by processing it and then calling #serve_with_file_origin 
  # on the controller. 
  #
  def serve(file, &block)
    markup    = read_markup(map_to_file(file))
    base_path = File.dirname(file)
    
    process(markup, base_path, &block)
  end
  
  def map_to_file(path)
    "#{path}.haml"
  end
    
  # Processes haml into html
  #
  def process(haml, base_path, &block)
    helper = Helper.new(base_path, self)

    Haml::Engine.new(haml).render(helper, {}, &block)
  end
end

lib/thin_ham/processors/markup.rb


# Generic markup processor that works by reading a file based on the 
# original filename, appending the markups extension. 
#
# This class is abstract, implementors should override #map_to_file and
# #process(markup), yielding target code. Using the Markup class itself
# will give you an identity transform, meaning delivering files as themselves.
#
# Example: 
#
#   class FooMarkup < Markup
#     def map_to_file(filename)
#       filename + ".foo"
#     end
#     def process(foo)
#       foo.gsub('foo', 'bar')
#     end
#   end
#
class ThinHam::Processors::Markup
  # Serves the file by processing it and then calling #serve_with_file_origin 
  # on the controller. 
  #
  def serve(file)
    markup = read_markup(map_to_file(file))
    
    process(markup)
  end
  
  # Reads markup file contents
  #
  def read_markup(file)
    File.read(file)
  end
  
  # Override these functions to implement a markup processor -----------------
  # (NOTE: Default implementation doesn't do anything and will leave files
  # unprocessed - this is interesting in its own right.)
  
  # Maps a path in the filesystem below the document root to a file that 
  # contains markup - returning the new path of the markup file. 
  #
  def map_to_file(path)
    path
  end
  
  # Processes markup into the target code. 
  #
  def process(markup)
    markup
  end
end

lib/thin_ham/hamlr.rb

require 'time'
require 'rack/mime'
require 'rack/utils'

require 'thin_ham/processors/markup'
require 'thin_ham/processors/sass'
require 'thin_ham/processors/haml'

# Serves files from the docroot given in the constructor. If a file 
# ends with the file ending of a known processor (.haml or .sass), transform
# the file and serve the result of the transformation. 
#
# This is intended to be a prototyping helper and thus it performs no caching. 
#
# CAUTION: Not for WAN exposed usage, not security reviewed !!! Sensible to 
# at least the '..'-kind of attack. 
#
# CAUTION: All files are currently loaded into memory and then served. This
# is probably not what you want to use for big movies.
#
# Based on rack/file.rb
#
class ThinHam::Hamlr
  attr_reader :docroot
  def initialize(docroot)
    @docroot = docroot
  end
  
  def call(env)
    dup.request(env)
  end
  
  # Handles request - this is thread-safe. 
  #
  def request(env)
    path_info = Rack::Utils.unescape(env["PATH_INFO"])
    full_path = File.join(docroot, path_info)

    processors.each do |processor|
      mapped_path = processor.map_to_file(full_path)
      if can_serve?(mapped_path)
        return http_answer(
          processor.serve(full_path), 
          mime_type(full_path))
      end
    end

    return not_found(path_info)
  end
  
  # Returns a list of processors known to man
  #
  def processors
    [
      ThinHam::Processors::Haml.new, 
      ThinHam::Processors::Sass.new, 
      ThinHam::Processors::Markup.new   # identity transform, serving files as files
    ]
  end
  
  # Can we serve this file? 
  #
  def can_serve?(file)
    File.file?(file) && File.readable?(file)
  end
  
  # Returns the mime type (a guess!) for the given +filename+
  #
  def mime_type(filename)
    Rack::Mime.mime_type(
      File.extname(filename), 
      'text/plain')
  end
  
  # Serves body given with mime_type and modif_date. 
  #
  def http_answer(body, mime_type)
    [200, {
      'Cache-Control'  => 'max-age=0',    # turns off browser cache (hopefully)
      "Content-Type"   => mime_type,
      "Content-Length" => body.size.to_s
    }, body]
  end
  
  # Returns a 404 not found answer. 
  #
  def not_found(path)
    body = "File not found: #{path}\n"
    [404, {
      'Cache-Control'  => 'max-age=3600',
      "Content-Type" => "text/plain",
      "Content-Length" => body.size.to_s
      },
      [body]]
  end
end